File: problem_solving_with_DSA/array/python/reverseWords.py
"""
Reverse words
"""
def _len(str):
    """ Lenght of a string
    Args:
        str (str): The string to find its length
    """
    length = 0
    for _ in str:
        length += 1
    return  length


def reverse(str, start, end):
    """ Reverse arary elements
    
    Args:
        str (str[chr]) - String to reverse
        start (int): First index of the string
        end (int): Last index of the string
    """
    word = ""
    
    for i in reversed(range(start, end+1)):
        word += str[i];
    
    return word
    

def reverseWords(str):
    """ Reverse individual words
    Args:
        str (str): The string to reverse its words
    """
    len = _len(str) 
    wstart = 0
    words = ""
    
    for i in range(len):
        if str[i] == ' ':
            words += reverse(str, wstart, i-1)
            words += ' '
            wstart = i + 1
    words += reverse(str, wstart, len-1)
    
    return words

if __name__ == '__main__':
    text = "Hello world"
    print(reverseWords(text))

File: problem_solving_with_DSA/array/C/reverseWords.c
#include <stdio.h>

/**
 * _strlen - Length of a string 
 * 
 * @str: String to find its length 
 * Return: its length 
 */
int _strlen(char str[])
{
    int len;
    for (len = 0; str[len] != '\0'; len++);
    return (len);
}

/**
 * reverse - Reverse String 
 * 
 * @str: String to reverse
 * @start: The start of the string
 * @end: The end of the string
 */
void reverse(char str[], int start, int end)
{
    while (start < end)
    {
        char temp = str[start];
        str[start++] = str[end];
        str[end--] = temp;
    }
}

/**
 * reverse_words - Reverse individual words 
 * 
 * @str: Words to reverse 
 */
void reverse_words(char str[])
{
    int len  = _strlen(str);
    int wstart = 0;
    int end =  0;
    
    for (int i = 0; i <= len; i++)
    {
        if (str[i] == ' ' || i == len)
        {
            reverse(str, wstart, end-1);
            wstart = i+1;
        }
        end++;
    }
}


/**
 * reverse_words2 - Reverse words 
 * 
 * @str: Words to reverse 
 */
void reverse_words2(char str[])
{
    int len  = _strlen(str);
    int wstart = 0;
    int end =  0;
    
    for (int i = 0; i < len; i++)
    {
        if (str[i] == ' ')
        {
            reverse(str, wstart, end-1);
            wstart = i+1;
        }
        end++;
    }
    reverse(str, wstart, len-1);
}

int main()
{
    char str[] = "Welcome to Log2Base2";
    
    printf("%s\n", str);
    reverse_words(str);
    printf("%s\n", str);
    
    return (0);
}

